### PR TITLE
Keep track of active status line messages to reduce message flickering

### PR DESCRIPTION
**Pull request by dennisschagt**
Fixes https://github.com/newsboat/newsboat/issues/238

TODO:
- [x] Write tests for `StatusLine` (I plan to do this later today but not sure if I find the time)

Created this PR to avoid duplicate work by other people; marked as draft as I still need to verify functionality using automatic tests.

### CODE REVIEW

**Review by Minoru — COMMENTED** [review]
This looks nice, thanks for working on this! I'll hold off approving until I see the tests though :)

Apart from tests, it'd be nice to have a bit of docs for this. While reading the code, the most obscure points to me were:

1. not immediately obvious how the three new classes hang together. A sentence like "StatusLine manages the array of messages, IStatus abstracts how messages are actually displayed, and AutoDiscardMessage is a RAII helper that removes its message when it's deconstructed" would be helpful;
2. in the cpp file, it took me a bit to realize that message ID 0 doesn't participate in deciding what to show — it's basically a toggle that just says "we're displaying a throwaway message that is not in the array of messages". I don't think it would be useful to explain how the array is managed—the code is clear enough—but I feel it's important to explain that ID 0 is special;
3. it'd be nice to have a comment explaining why we don't care about `next_message_id` wrapping back to 0.

File: src/statusline.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+std::shared_ptr<AutoDiscardMessage> StatusLine::show_message_until_finished(
+	const std::string& message)
+{
+	std::lock_guard<std::mutex> guard(m);
+
+	if (next_message_id == 0) {
+		next_message_id++;
+	}
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply src/statusline.cpp:40]: I wrote it like this to guarantee that ID 0 is never used, even in the case of a (highly ulikely) wrap around of `next_message_id`.

**Comment by dennisschagt** [comment]
Thank you for the review and the clear suggestions!

File: src/statusline.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Minoru) [reply src/statusline.cpp:40]: Ah, makes sense, thank you! I see you've added a comment about this, so consider this resolved.

File: test/statusline.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	{
+		last_message = message;
+		last_string_was_error = true;
+	}
+
+	std::string last_message;
+	bool last_string_was_error;
+};
+
+TEST_CASE("StatusLine passes status messages to registered IStatus implementation")
```
Comment (Minoru) [inline test/statusline.cpp:28]: Please add a second argument, "[StatusLine]", to this and the next test cases. That way, you can run *just* them with `./test/test '[StatusLine]'`, which can be helpful during debugging.

File: test/statusline.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+		const auto previous_message_lifetime =
+			status_line.show_message_until_finished("previous message...");
+
+		WHEN("show_message() is called") {
+			status_line.show_message("this is a status");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "this is a status");
+				REQUIRE_FALSE(status_message_handler.last_string_was_error);
+			}
```
Comment (Minoru) [inline test/statusline.cpp:122]: I think this should also check what happens if `previous_message_lifetime` ends after a call to `show_message()`.

File: test/statusline.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+		}
+
+		WHEN("show_error() is called") {
+			status_line.show_error("this is an error");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "this is an error");
+				REQUIRE(status_message_handler.last_string_was_error);
+			}
```
Comment (Minoru) [inline test/statusline.cpp:131]: Same as above: I think this should also check what happens if `previous_message_lifetime` ends after a call to `show_error()`.

File: test/statusline.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+		WHEN("the latest message goes out of scope") {
+			message_lifetime3.reset();
+
+			THEN("the most recent active status message is restored") {
+				REQUIRE(status_message_handler.last_message == "second message");
+
+				THEN("a further destruction causes the then latest message to be restored") {
+					message_lifetime2.reset();
+					REQUIRE(status_message_handler.last_message == "first message");
```
Comment (Minoru) [inline test/statusline.cpp:175]: I'd add one final check to see what happens when the last remaining lifetime ends.

**Review by Minoru — APPROVED** [review]
I've got a couple requests regarding tests. Other than that, this looks excellent! Feel free to merge once you addressed my feedback.

Thanks once again for working on this!

**Comment by dennisschagt** [comment]
@dependabot merge

Just checking if this works a non-dependency PR

**Comment by dennisschagt** [comment]
@Minoru Good points, all taken care of. Thanks for the reviews!

**Comment by Minoru** [comment]
> Just checking if this works a non-dependency PR

That reminds me… I just enabled auto-merges for this repo, so you can [try choosing the auto-merge option from a drop-down on the merge button](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request). Let's see how well that works :)

**Pull request merged**